src/app/store/keys/key.actions.ts

```diff
@@ -14,31 +17,47 @@ import { inMemoryKeySlice } from '../in-memory-key/in-memory-key.slice';
 import { selectCurrentKey } from './key.selectors';
 import { defaultKeyId, keySlice } from './key.slice';
 
-async function restoredWalletHighestGeneratedAccountIndex(secretKey: string) {
-  try {
-    // Don't need, nor return, encrypted wallet value, so a legit password isn't
-    // needed. Ideally `@stacks/wallet-sdk` should be updated so that the encrypt
-    // function is a separate method
-    const wallet = await generateWallet({ secretKey, password: '' });
-    // use network to select addresses based on owned usernames
-    const network = new StacksMainnet();
-    const restoredWallet = await restoreWalletAccounts({
-      wallet,
-      gaiaHubUrl: gaiaUrl,
-      network,
-    });
-    return restoredWallet.accounts.length - 1;
-  } catch (e) {
-    return 0;
-  }
-}
+function setWalletEncryptionPassword(args: { password: string; client: StacksClient }): AppThunk {
+  const { password, client } = args;
 
-const setWalletEncryptionPassword = (password: string): AppThunk => {
   return async (dispatch, getState) => {
     const secretKey = selectDefaultWalletKey(getState());
     if (!secretKey) throw new Error('Cannot generate wallet without first having generated a key');
     const { encryptedSecretKey, salt } = await encryptMnemonic({ secretKey, password });
-    const highestAccountIndex = await restoredWalletHighestGeneratedAccountIndex(secretKey);
+
+    const legacyAccountActivityLookup =
+      await checkForLegacyGaiaConfigWithKnownGeneratedAccountIndex(secretKey);
+
+    async function doesStacksAddressHaveBalance(address: string) {
+      const resp = await client.accountsApi.getAccountBalance({ principal: address });
+      return Number(resp.stx.balance) > 0;
+    }
+
+    async function doesStacksAddressHaveBnsName(address: string) {
+      const resp = await client.namesApi.getNamesOwnedByAddress({ address, blockchain: 'stacks' });
+      return resp.names.length > 0;
+    }
+
+    // Performs a recursive check for account activity. When activity is found
+    // at a higher index than what is found on Gaia (long-term wallet users), we
+    // update the highest known account index that the wallet generates. This
+    // action is performed outside this Promise's execution, as it may be slow,
+    // and the user shouldn't have to wait before being directed to homepage.
+    void recurseAccountsForActivity({
+      async doesAddressHaveActivityFn(index) {
+        const address = getStacksAddressByIndex(secretKey, AddressVersion.MainnetSingleSig)(index);
+        const hasBal = await doesStacksAddressHaveBalance(address);
+        const hasNames = await doesStacksAddressHaveBnsName(address);
+        // for demo purposes, remove
+        // eslint-disable-next-line no-console
+        console.log({ index, address, hasBal, hasNames });
```

Comment:
Are you intending to remove this before merging?

Reply:
removec